packages/cosmic-swingset/lib/ag-solo/vats/chain-config.json

```diff
@@ -0,0 +1,49 @@
+{
+  "bootstrap": "bootstrap",
+  "bundles": {
+    "zcf": {
+      "sourcePath": "../../../../zoe/src/contractFacet.js"
+    }
+  },
+  "vats": {
+    "board": {
+      "sourcePath": "vat-board.js"
```

Comment:
Relevant to my comment on how to resolve these things, I would suggest changing `"sourcePath"` to `"sourceSpecifier"` and rewrite these like `"./vat-board.js"`.  There's good reasons to be compatible with ESM resolution semantics.

Reply:
`sourceSpecifier` actually makes a more sense if the thing might not actually be a path per se, which it soon won't be.

Reply:
I'm going to make this change, but if you don't mind I'd like to put it in its own PR as it's going to touch a lot of files that are not relevant to the changes in this PR.

I'm quite annoyed that GitHub does not seem to have very good workflow around change requests in PRs.

Reply:
I guess I'd be happiest about Zoe using `"exports"` to explicitly offer that file, medium-happy about using the deep-import of `@zoe/src/contractFacet.js` without `"exports"` (revealing the internal layout of zoe but letting us add `exports` in the future if we ever wanted to move it without a compatiiblity break, and least happy about the "up and over" `../../../../../../../../../../omg/where/am/i/now/zoe/something` path.

Monorepos let us (mostly) abandon the practice of making new releases just to allow changes from one package become visible to some other package, increasing velocity. They're kind of downstream-hostile, because without well-considered releases, we don't really pay attention to what a downstream/outside-consumer observes: they can't use new changes at all until we happen to make a release, and then the difference from one version (of e.g. swingset) to the next is kind of random. Conventional-commits increases the chances that the resulting semver is increased appropriately, which is good.

If we allow cross-package filepaths into our code, then e.g. cosmic-swingset simply won't work *at all* outside the monorepo, and we've given up all pretense of making that package usable by itself. We shouldn't be [publishing](https://www.npmjs.com/package/@agoric/cosmic-swingset) such a package to NPM: it's useless, there's no practical way to install it in a place where that `../../` path can possibly point to the right place.

(Note: I'm in favor of `readConfigFile()` *supporting* arbitrary pathnames like that, interpreted relative to the JSON config file. I just don't want us to use them)

My previous concerns still apply, about `@agoric/zoe/src/zcf.js` being interpreted relative to *swingset*'s `node_modules`/`dependencies`, rather than relative to the importing application (cosmic-swingset). I think that would continue to work in an NPM-installed cosmic-swingset directory: cosmic-swingset will declare a dependency on zoe, or will declare a dependency on an `@agoric/ecentral-services` package which then pulls in zoe, and unless there are incompatible versions being referenced, everything should be put into the same top-level `node_modules/` directory, where swingset will happen to look for them. Not great, not precise, but it'll probably work until/if Endo has something different to say about it.

My most preferred solution is still to have Zoe export a new function, which returns the pathname of the zcf.js file. This makes it clear that Zoe's ZCF support has two pieces that must match: the contents of the ZCF bundle, and the name that Zoe will use to reference the bundle when it makes a new vat. The Zoe documentation would say "to use zoe, your swingset configuration needs to cite vat-zoe in this way, it needs to call this function to get the ZCF pathname, it needs to arrange for that file to be bundled somehow, and it needs to tell zoe about the resulting bundle name via the vatParameters".

In a perfect world, the zoe vat-making code would just have a special magical `import`-like statement which sets up the association between name and contents (just like regular `import` does, except deferred until runtime and placed in a new vat), and there'd be some magic way for cosmic-swingset to learn enough about this mapping to tell swingset about it. Or for swingset itself to notice that it's loading a vat which has a marker like this in the code, some sort of auxilliary bundle syntax noticed by `bundleSource`. The Zoe docs wouldn't have to explain this to the host author at all. But I don't know what that magic would be yet.

My second most preferred solution is to remove the `bundles` section from the JSON file, and have cosmic-swingset's setup code run `const config = loadConfigFile(..); config.bundles.zcf = require.resolve('@agoric/zoe/src/zcf.whatever');`, which gets the path name relative to cosmic-swingset's `node_modules/`. The zoe docs explain the host requirements in a different way, and zoe commits to maintaining that particular exported pathname (but doesn't add a new exported function that deals in pathnames).

But I know it's kinda gross to get 95% of the config from a file and then have to run code to fill in that one remaining bit. So my third most preferred solution is to implement @michaelfig 's `sourceSpecifier` scheme (I think that's a good idea regardless), and have `chain-config.json` cite `@agoric/zoe/src/zcf.js`. That confuses who knows about Zoe, but it'll probably work in all the environments I can think of. The Zoe documentation is mostly the same as in the second case.